### PR TITLE
Fixes for CDAP-6115, CDAP-6121, CDAP-6206

### PR DESCRIPTION
**Pull request by chtyim**
- (CDAP-6115) Handle both primitive and boxed type upscaling
- (CDAP-6121) Allow @UseDataset usage in Spark init and destroy
  - Also did a cleanup of existing configurer code
- (CDAP-6206) Improve error message if Spark is missing

### CODE REVIEW

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT4600-2

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java [hunk 1]
```diff
@@ -148,7 +149,42 @@ private ConfigResponse createResponse(Application app)
...
+          // If the missing class has "spark" in the name, try to see if Spark is available
+          if (missingClass.startsWith("org.apache.spark.") || missingClass.startsWith("co.cask.cdap.api.spark.")) {
+            File sparkAssemblyJar;
+            try {
+              sparkAssemblyJar = SparkUtils.locateSparkAssemblyJar();
+            } catch (Exception e) {
+              // Spark is not available, it is most likely caused by missing Spark in the platform
+              throw new IllegalStateException(
+                "Missing Spark related class " + missingClass +
+                  ". It may be caused by unavailability of Spark in the platform.", t);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java]: "platform" is a little hard to understand for users. Is what you mean that Spark is not installed in the cluster? Or that SPARK_HOME is not set? Perhaps it is not set in cdap-env.sh?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java [hunk 2]
```diff
@@ -148,7 +149,42 @@ private ConfigResponse createResponse(Application app)
...
+
+            // Spark is available, can be caused by incompatible Spark version
+            throw new InvalidArtifactException(
+              "Missing Spark related class " + missingClass +
+                ". Configured to use Spark assembly jar located at " + sparkAssemblyJar +
+                ", which may be incompatible with the one required by the application", t);
+
+          }
+          // If Spark is available or the missing class is not a spark related class,
+          // then the missing class is most lifely due to some missing library in the artifact jar
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java]: likely

File: cdap-common/src/main/java/co/cask/cdap/internal/io/DatumWriterGenerator.java [hunk 3]
```diff
@@ -404,9 +404,15 @@ private void encodeSimple(GeneratorAdapter mg, TypeToken<?> type, Schema schema,
     TypeToken<?> encodeType = type;
     mg.loadArg(encoder);
     mg.loadArg(value);
-    if (Primitives.isWrapperType(encodeType.getRawType())) {
+
+    boolean isPrimitiveBoxType = Primitives.isWrapperType(encodeType.getRawType());
+    if (encodeType.getRawType().isPrimitive() || isPrimitiveBoxType) {
+      // Unwrap it if it is boxed type, otherwise it won't change the encodeType.
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/internal/io/DatumWriterGenerator.java:410]: nit: does this comment belong on line 412 instead?

**Comment by bdmogal** [comment]
tiny nit. Otherwise 💯

File: cdap-common/src/main/java/co/cask/cdap/internal/io/DatumWriterGenerator.java (hunk 3 above)
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/internal/io/DatumWriterGenerator.java:410]: You mean line 411? 412 is an empty line. It is, that's why it's one line before it.

**Comment by chtyim** [comment]
Build passed again. http://builds.cask.co/browse/CDAP-DUT4600-4

File: cdap-common/src/main/java/co/cask/cdap/internal/io/DatumWriterGenerator.java (hunk 3 above)
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/internal/io/DatumWriterGenerator.java:410]: I meant before line 413. But I missed that you have `|| isPrimitiveBoxType` on line 409 as well. So its probably ok where it is.

Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/internal/io/DatumWriterGenerator.java:410]: The comment is intended for the unwrapping of boxed type to primitive type. Line 413 is for generating an unbox statement in the bytecode.

**Pull request merged**